futhark_ffi/__init__.py

```diff
@@ -171,6 +185,34 @@ def wrapper(*args):
 
         return wrapper
 
+    def make_store_wrapper(self, ff):
+        
+        @wraps(ff)
+        def wrapper(opaque):
+            bytes_ptr_ptr = self.ffi.new('void **', self.ffi.NULL)
+            size_ptr = self.ffi.new('size_t *')
+            err = ff(self.ctx, opaque, bytes_ptr_ptr, size_ptr)
+            self._errorcheck(err)
+            bytes_ptr = self.ffi.gc(bytes_ptr_ptr[0], self.lib.free)
+            return self.ffi.buffer(bytes_ptr, size_ptr[0])
+
+        return wrapper
+    
+    def make_restore_wrapper(self, ff):
+        fut_type = self.ffi.typeof(ff).result
+        
+        @wraps(ff)
+        def wrapper(buffer):
+            bytes_ptr = self.ffi.from_buffer(buffer)
+            res = ff(self.ctx, bytes_ptr)
+            #handle NULL pointer
+            if res:
+                return self.ffi.gc(res, partial(self.types[fut_type].free, self.ctx))
+            else:
+                raise ValueError("failed to restore value from buffer")
```

Comment:
Should we not retrieve an error message from the context in this case?

Reply:
Oh, I wasn't sure if futhark gave errors here since the mechanism is different (Null pointer instead of error code), I will add it.

Reply:
I tried this, but it didn't raise an error.
```
            if res:
                return self.ffi.gc(res, partial(self.types[fut_type].free, self.ctx))
            else:
                errptr = self.lib.futhark_context_get_error(self.ctx)
                if errptr:
                    raise ValueError(self._get_string(errptr))
```

It appears that `errptr` is `NULL` here, as I was suspecting, or am I missing something?

Reply:
Oh well, then stick to the original version.